In [3]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Lasso, Ridge
from sklearn.svm import SVR
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.decomposition import PCA
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

In [4]:
df = pd.read_csv("Hyderabad.csv")
df

,Price,Area,Location,No. of Bedrooms,Resale,MaintenanceStaff,Gymnasium,SwimmingPool,LandscapedGardens,JoggingTrack,...,LiftAvailable,BED,VaastuCompliant,Microwave,GolfCourse,TV,DiningTable,Sofa,Wardrobe,Stadium
0,6968000,1340,Nizampet,2,0,0,1,1,1,1,...,1,0,1,0,0,0,0,0,0,0
1,29000000,3498,Hitech City,4,0,0,1,1,1,1,...,1,0,1,0,0,0,0,0,0,0
2,6590000,1318,Manikonda,2,0,0,1,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3,5739000,1295,Alwal,3,1,0,0,0,0,0,...,1,0,0,0,0,0,0,0,0,0
4,5679000,1145,Kukatpally,2,0,0,0,0,1,0,...,1,0,0,0,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2513,11000000,1460,Nacharam,2,1,9,9,9,9,9,...,9,9,9,9,9,9,9,9,9,9
2514,26000000,1314,Manikonda,2,1,9,9,9,9,9,...,9,9,9,9,9,9,9,9,9,9
2515,13300000,2625,Madhapur,3,1,9,9,9,9,9,...,9,9,9,9,9,9,9,9,9,9
2516,10800000,2050,Hitech City,3,0,9,9,9,9,9,...,9,9,9,9,9,9,9,9,9,9


In [5]:
df.shape

(2518, 40)

In [6]:
df = df.drop_duplicates()

In [7]:
df.shape

(1998, 40)

In [8]:
df.isnull().sum()

Price                  0
Area                   0
Location               0
No. of Bedrooms        0
Resale                 0
MaintenanceStaff       0
Gymnasium              0
SwimmingPool           0
LandscapedGardens      0
JoggingTrack           0
RainWaterHarvesting    0
IndoorGames            0
ShoppingMall           0
Intercom               0
SportsFacility         0
ATM                    0
ClubHouse              0
School                 0
24X7Security           0
PowerBackup            0
CarParking             0
StaffQuarter           0
Cafeteria              0
MultipurposeRoom       0
Hospital               0
WashingMachine         0
Gasconnection          0
AC                     0
Wifi                   0
Children'splayarea     0
LiftAvailable          0
BED                    0
VaastuCompliant        0
Microwave              0
GolfCourse             0
TV                     0
DiningTable            0
Sofa                   0
Wardrobe               0
Stadium                0


In [9]:
df['Location'] = LabelEncoder().fit_transform(df['Location'])

C:\Users\Zinkal Patel\AppData\Local\Temp\ipykernel_7196\4176234621.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Location'] = LabelEncoder().fit_transform(df['Location'])


In [10]:
df.head()

,Price,Area,Location,No. of Bedrooms,Resale,MaintenanceStaff,Gymnasium,SwimmingPool,LandscapedGardens,JoggingTrack,...,LiftAvailable,BED,VaastuCompliant,Microwave,GolfCourse,TV,DiningTable,Sofa,Wardrobe,Stadium
0,6968000,1340,162,2,0,0,1,1,1,1,...,1,0,1,0,0,0,0,0,0,0
1,29000000,3498,85,4,0,0,1,1,1,1,...,1,0,1,0,0,0,0,0,0,0
2,6590000,1318,132,2,0,0,1,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3,5739000,1295,9,3,1,0,0,0,0,0,...,1,0,0,0,0,0,0,0,0,0
4,5679000,1145,118,2,0,0,0,0,1,0,...,1,0,0,0,0,0,0,0,0,0


In [17]:
X = df.drop('Price', axis=1)
y = df['Price']


In [18]:
Xtrain, Xtest, ytrain, ytest = train_test_split(X, y, test_size=0.2, random_state=42)

In [19]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(Xtrain)
X_test_scaled = scaler.transform(Xtest)

In [20]:
models = {
    "Simple Linear Regression": LinearRegression(),
    "Multiple Linear Regression": LinearRegression(),
    "Multiple Linear Regression with PCA": Pipeline([
            ('pca', PCA(n_components=10)),
           ('lr', LinearRegression())]),
    "Lasso Regression": Lasso(alpha=1.0),
    "Ridge Regression": Ridge(alpha=1.0),
    "SVR": SVR(kernel='rbf', C=100, gamma=0.1, epsilon=.1),
    "Decision Tree": DecisionTreeRegressor(max_depth=5, random_state=42),
    "Random Forest": RandomForestRegressor(n_estimators=100, random_state=42)
}

In [21]:
results = []

for name, model in models.items():
    if name == "Simple Linear Regression":
        model.fit(X_train_scaled[:, 0].reshape(-1, 1), ytrain)
        ypred = model.predict(X_test_scaled[:, 0].reshape(-1, 1))
    else:
        model.fit(X_train_scaled, ytrain)
        ypred = model.predict(X_test_scaled)
    
    mae = mean_absolute_error(ytest, ypred)
    mse = mean_squared_error(ytest, ypred)
    rmse = np.sqrt(mse)
    r2 = r2_score(ytest, ypred)
    
    results.append({
        "Model": name,
        "MAE": mae,
        "MSE": mse,
        "RMSE": rmse,
        "R²": r2
    })


D:\ANACONDA\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:697: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 2.340e+16, tolerance: 1.458e+13
  model = cd_fast.enet_coordinate_descent(


In [23]:
print(results)

[{'Model': 'Simple Linear Regression', 'MAE': 2403104.1712970287, 'MSE': 16510022685278.074, 'RMSE': 4063252.7222999646, 'R²': 0.7223135333245828}, {'Model': 'Multiple Linear Regression', 'MAE': 2274380.316201901, 'MSE': 15187611732858.342, 'RMSE': 3897128.652336017, 'R²': 0.7445555151722367}, {'Model': 'Multiple Linear Regression with PCA', 'MAE': 2290185.582268053, 'MSE': 15343368769333.213, 'RMSE': 3917061.2414580924, 'R²': 0.7419357961116986}, {'Model': 'Lasso Regression', 'MAE': 2274184.808630807, 'MSE': 15166249588364.05, 'RMSE': 3894386.933570424, 'R²': 0.7449148107673009}, {'Model': 'Ridge Regression', 'MAE': 2265992.4274554667, 'MSE': 15170103349469.977, 'RMSE': 3894881.686196639, 'R²': 0.7448499933333508}, {'Model': 'SVR', 'MAE': 4379510.184270227, 'MSE': 61327832492140.71, 'RMSE': 7831208.878081385, 'R²': -0.03148914076235054}, {'Model': 'Decision Tree', 'MAE': 2226058.806720136, 'MSE': 15374336084443.24, 'RMSE': 3921012.1250058943, 'R²': 0.7414149485950567}, {'Model': 'Rand

In [24]:
results_df = pd.DataFrame(results)
print(results_df)

                                 Model           MAE           MSE  \
0             Simple Linear Regression  2.403104e+06  1.651002e+13   
1           Multiple Linear Regression  2.274380e+06  1.518761e+13   
2  Multiple Linear Regression with PCA  2.290186e+06  1.534337e+13   
3                     Lasso Regression  2.274185e+06  1.516625e+13   
4                     Ridge Regression  2.265992e+06  1.517010e+13   
5                                  SVR  4.379510e+06  6.132783e+13   
6                        Decision Tree  2.226059e+06  1.537434e+13   
7                        Random Forest  1.770816e+06  1.195837e+13   

           RMSE        R²  
0  4.063253e+06  0.722314  
1  3.897129e+06  0.744556  
2  3.917061e+06  0.741936  
3  3.894387e+06  0.744915  
4  3.894882e+06  0.744850  
5  7.831209e+06 -0.031489  
6  3.921012e+06  0.741415  
7  3.458088e+06  0.798869  


Conclusion

From the results:


Random Forest Regressor is the best fit for your Hyderabad dataset because it achieves the lowest prediction errors and highest R² score. It can model complex, nonlinear relationships between features, making it more effective for real-world property price prediction than linear or regularized models.

Key Points:

Lowest Errors – Random Forest has the smallest MAE (1.77M) and RMSE (3.46M) among all models, meaning it predicts house prices more accurately than others.

Highest R² Score – At 0.799, it explains ~80% of the variation in the target variable, better than the ~74% achieved by linear and ridge regression.

Nonlinear Relationships – Real estate pricing often has nonlinear interactions between features (location, size, amenities), which Random Forest captures well, unlike simple linear models.

Robustness – It handles outliers and irrelevant features better without heavy preprocessing.